internal/provider/data_source_gitlab_repository_tree.go

```diff
@@ -0,0 +1,135 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/mitchellh/hashstructure"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+var _ = registerDataSource("gitlab_repository_tree", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_repository_tree`" + ` data source allows details of directories and files in a repository to be retrieved.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/repositories.html#list-repository-tree)`,
+
+		ReadContext: dataSourceGitlabRepositoryTreeRead,
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description: "The ID or full path of the project owned by the authenticated user.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"path": {
+				Description: "The path inside repository. Used to get content of subdirectories.",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"ref": {
+				Description: "The name of a repository branch or tag.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
```

Comment:
NIT: Let's define the required attributes first:

```suggestion
			"ref": {
				Description: "The name of a repository branch or tag.",
				Type:        schema.TypeString,
				Required:    true,
			},
			"path": {
				Description: "The path inside repository. Used to get content of subdirectories.",
				Type:        schema.TypeString,
				Optional:    true,
			},
```

Reply:
Reordered these and also did the same for the repo file resource/datasource